homeassistant/components/romy/vacuum.py

```diff
@@ -0,0 +1,132 @@
+
+from homeassistant.components.vacuum import StateVacuumEntity, VacuumEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_NAME, ATTR_SW_VERSION
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN, LOGGER
+from .coordinator import RomyVacuumCoordinator
+
+ICON = "mdi:robot-vacuum"
+
+FAN_SPEED_NONE = "default"
+FAN_SPEED_NORMAL = "normal"
+FAN_SPEED_SILENT = "silent"
+FAN_SPEED_INTENSIVE = "intensive"
+FAN_SPEED_SUPER_SILENT = "super_silent"
+FAN_SPEED_HIGH = "high"
+FAN_SPEED_AUTO = "auto"
+
+FAN_SPEEDS: list[str] = [
+    FAN_SPEED_NONE,
+    FAN_SPEED_NORMAL,
+    FAN_SPEED_SILENT,
+    FAN_SPEED_INTENSIVE,
+    FAN_SPEED_SUPER_SILENT,
+    FAN_SPEED_HIGH,
+    FAN_SPEED_AUTO,
+]
+
+# Commonly supported features
+SUPPORT_ROMY_ROBOT = (
+    VacuumEntityFeature.BATTERY
+    | VacuumEntityFeature.PAUSE
+    | VacuumEntityFeature.RETURN_HOME
+    | VacuumEntityFeature.STATE
+    | VacuumEntityFeature.START
+    | VacuumEntityFeature.STOP
+    | VacuumEntityFeature.FAN_SPEED
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up ROMY vacuum cleaner."""
+
+    coordinator: RomyVacuumCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+    romy: RomyRobot = coordinator.romy
+
+    romy_vacuum_entity = RomyVacuumEntity(coordinator, romy)
+
+    entities = [romy_vacuum_entity]
+    async_add_entities(entities, True)
+
+
+class RomyVacuumEntity(CoordinatorEntity[RomyVacuumCoordinator], StateVacuumEntity):
+    """Representation of a ROMY vacuum cleaner robot."""
+
+    _attr_has_entity_name = True
+    _attr_name = None
+    _attr_supported_features = SUPPORT_ROMY_ROBOT
+    _attr_fan_speed_list = FAN_SPEEDS
+    _attr_icon = ICON
+
+    def __init__(
+        self,
+        coordinator: RomyVacuumCoordinator,
+        romy: RomyRobot,
+    ) -> None:
+        """Initialize the ROMY Robot."""
+        super().__init__(coordinator)
+        self.romy = romy
+        self._attr_unique_id = self.romy.unique_id
+        self._device_info = DeviceInfo(
+            identifiers={(DOMAIN, romy.unique_id)},
+            manufacturer="ROMY",
+            name=romy.name,
+            model=romy.model,
+            sw_version=romy.firmware,
+        )
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._attr_fan_speed = FAN_SPEEDS[self.romy.fan_speed]
+        self._attr_battery_level = self.romy.battery_level
+        self._attr_state = self.romy.status
+
+        self._device_info[ATTR_NAME] = self.romy.name
+        self._device_info[ATTR_SW_VERSION] = self.romy.firmware
```

Comment:
My impression is that device info is read at startup when registering the entity/device. I don't think it is expected that it can just change like this. (If there is a specific use case in mind here,  we can discuss...  I've never seen code update device info like this -- and if it were to update, it would probably need to happen through the device registry).

I would recommend just not doing this here.

Reply:
removed